app/services/SalesforceService.scala

```diff
@@ -0,0 +1,99 @@
+package services
+
+import akka.agent.Agent
+import com.gu.membership.salesforce.Member.Keys
+import com.gu.membership.salesforce.{Authentication, MemberId, MemberRepository, Scalaforce}
+import com.typesafe.scalalogging.LazyLogging
+import configuration.Config
+import model.PersonalData
+import play.api.Logger
+import play.api.libs.concurrent.Akka
+import play.api.libs.json.{JsObject, Json}
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration.{FiniteDuration, _}
+import scala.concurrent.{Await, Future}
+
+trait SalesforceService extends LazyLogging {
+
+  def createOrUpdateUser(personalData: PersonalData, userId: UserId): Future[Option[MemberId]]
+
+  def createSalesforceUserData(personalData: PersonalData): JsObject = {
+    Seq(Json.obj(
+      Keys.EMAIL -> personalData.email,
+      Keys.FIRST_NAME -> personalData.firstName,
+      Keys.LAST_NAME -> personalData.lastName,
+      Keys.MAILING_STREET -> personalData.address.street,
+      Keys.MAILING_CITY -> personalData.address.town,
+      Keys.MAILING_POSTCODE -> personalData.address.postcode,
+      Keys.MAILING_COUNTRY -> "United Kingdom",
+      Keys.ALLOW_MEMBERSHIP_MAIL -> true
+    )) ++ Map(
+      Keys.ALLOW_THIRD_PARTY_EMAIL -> Some(true),
+      Keys.ALLOW_GU_RELATED_MAIL -> Some(true)
+    ).collect { case (k, Some(v)) => Json.obj(k -> v) }
+  }.reduce(_ ++ _)
+
+}
+
+object SalesforceService extends SalesforceService {
+  override def createOrUpdateUser(personalData: PersonalData, userId: UserId): Future[Option[MemberId]] =
+    SalesforceRepo.upsert(userId.id, createSalesforceUserData(personalData)).map(Some(_))
+}
+
+object SalesforceRepo extends MemberRepository {
+  override val salesforce = new Scalaforce {
+    override val consumerKey = Config.Salesforce.consumerKey
+    override val apiUsername = Config.Salesforce.apiUsername
+    override val consumerSecret = Config.Salesforce.consumerSecret
+    override val apiToken = Config.Salesforce.apiToken
+    override val apiPassword = Config.Salesforce.apiPassword
+    override val application = Config.appName
+    override val apiURL = Config.Salesforce.apiURL.toString
+    override val stage = Config.Salesforce.envName
+
+    val authTask = ScheduledTask("", Authentication("", ""), 0.seconds, 30.minutes)(getAuthentication)
+
+    def authentication: Authentication = authTask.get()
+  }
+}
+
+trait ScheduledTask[T] {
+  import play.api.Play.current
+
+  val initialValue: T
+  val initialDelay: FiniteDuration
+  val interval: FiniteDuration
+
+  val name = getClass.getSimpleName
+
+  private implicit val system = Akka.system
+  lazy val agent = Agent[T](initialValue)
+
+  def refresh(): Future[T]
```

Comment:
Does `refresh` need to be a `Future[T]`. `agent.sendOff` is [executed asynchronously](http://doc.akka.io/docs/akka/snapshot/scala/agents.html), so maybe I would get rid of the Future wrapper type and of the `Await.result` statement:

```scala
def refresh: T
...
agent.sendOff { _ => refresh }
```